core/src/main/java/com/linecorp/armeria/server/DefaultDependencyInjector.java

```diff
@@ -0,0 +1,118 @@
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linecorp.armeria.server;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Supplier;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.MoreObjects;
+
+import com.linecorp.armeria.common.annotation.Nullable;
+
+final class DefaultDependencyInjector implements DependencyInjector {
+
+    private static final Logger logger = LoggerFactory.getLogger(DefaultDependencyInjector.class);
+
+    private final Map<Class<?>, Object> singletons;
+    private final Map<Class<?>, Supplier<?>> singletonSuppliers;
+    private final Map<Class<?>, Supplier<?>> prototypes;
+    private final List<AutoCloseable> closeablePrototypes = new ArrayList<>();
+
+    DefaultDependencyInjector(Map<Class<?>, Object> singletons,
+                              Map<Class<?>, Supplier<?>> singletonSuppliers,
+                              Map<Class<?>, Supplier<?>> prototypes) {
+        this.singletons = new HashMap<>(singletons.size() + singletonSuppliers.size());
+        this.singletons.putAll(singletons);
+        this.singletonSuppliers = singletonSuppliers;
+        this.prototypes = prototypes;
+    }
+
+    @Override
+    public synchronized <T> T getInstance(Class<T> type) {
+        final Object instance = singletons.get(type);
+        if (instance != null) {
+            //noinspection unchecked
+            return (T) instance;
+        }
+        T supplied = getInstance(type, singletonSuppliers);
+        if (supplied != null) {
+            singletons.put(type, supplied);
+            return supplied;
+        }
+
+        supplied = getInstance(type, prototypes);
+        if (supplied instanceof AutoCloseable) {
+            closeablePrototypes.add((AutoCloseable) supplied);
+        }
+
+        return supplied;
+    }
+
+    @Nullable
+    private static <T> T getInstance(Class<T> type, Map<Class<?>, Supplier<?>> suppliers) {
+        final Supplier<?> supplier = suppliers.get(type);
+        if (supplier != null) {
+            final Object supplied = supplier.get();
+            requireNonNull(supplied, supplier + " returns null.");
+            if (!type.isInstance(supplied)) {
+                throw new IllegalArgumentException(supplied + " is not an instance of " + type.getName());
+            }
+            //noinspection unchecked
+            return (T) supplied;
+        }
+        return null;
+    }
+
+    @Override
+    public synchronized void close() {
```

Comment:
It looks like a caller can still call `getInstance()` after `close()` is called. Is this intended?

Reply:
At first, that was my intention. But it does not work that way anymore so let me change the implementation. Thanks!